### ISSUE TITLE
Python 3.13 support 

### ISSUE THREAD

**Issue by marcellodesales** [Nuitka/Nuitka#2834]
Here we are going to track the progress as it happens for Python 3.13 support. For the first time since maybe 3.5, we will be aiming at support in a release when CPython 3.13 is final.

**Comment by kayhayen** [comment]
Yeah, let me state my plan here. With 3.12 completed, I want to aim at supporting 3.13 at the release date, together with a no-GIL demonstration of C integer (or float) loops running at C speed on multiple threads. So, I think of this one as big priority.

At this time I have no idea, how difficult no-GIL is going to be for Nuitka. Surely no all code ever written or generated for it, is safe against races, although generally we should never release values too soon, that might be the case.

Normally I would wait for the release candidates, but this time I might even look into beta versions. I have been burned before, implementing beta features that later got removed or reworken, but I don't expect the no-GIL to be one of those at this point.

**Comment by kayhayen** [comment]
So, 3.12 support is completed and will be released with 2.3. I have downloaded and installed Python 3.13beta1 and resolved the first issues with the C header changes needed for that version, so it compiles now to C where it is giving a bunch of errors.

I will try and tackle first the apparent removal of the "_PyCFrame" as "cframe" in the interpreter state. This was always strange to me, ever since it was introduced in 3.11 and Nuitka honors it, but I didn't get why it exists. The frame pushing and poping codes for functions, generators, etc. will have to see this abstracted. 

Like I said, I will aim to support Python 3.13 as early as possible. At this time, I think I want "Hello World" with beta 1 to work in the remainder of the month, but it's not yet a priority.

**Comment by kayhayen** [comment]
So, the frame abstraction is done, that was easy. Only the function to dump the frame stack has not yet been adapted, as it goes more into the C frames there, and probably needs more though, but it might be a debug only code. I saw no errors from it immediately.

I noticed that the way freelist states are stored has changed, and I adapted it for slices. While doing it, I noticed that we were getting the interpreter state, which can be an API call. Now it's in the state, and that's being passed now, as it's already done with many code helpers. Then it's easily abstracted.

```
#if PYTHON_VERSION >= 0x3d0
#ifdef Py_GIL_DISABLED
    PySliceObject **slice_cache_ptr = &((_PyThreadStateImpl *)tstate)->freelists.slices.slice_cache;
#else
    PySliceObject **slice_cache_ptr = &tstate->interp->object_state.freelists.slices.slice_cache;
#endif
#else
    PyInterpreterState *interp = tstate->interp;
    PySliceObject **slice_cache_ptr = &interp->slice_cache;
#endif
```

This is also the first time I saw a distinction between GIL and no-GIL builds. I will focus on the GIL build first since that will be the ground base, but the goal is to get no-GIL to work as well. Of course, the main goal of supporting 3.13 early is to be able to show it off as soon as possible.

The passing of `tstate` should actually make things faster. I noticed that in order to avoid having to lose referenced passed in or not, something Nuitka does a lot, e.g., when constructing tuples, CPython now has variants that consume references or not for slice creation. I am very tempted to generate dedicated slice creation helpers, too, now and make use of them. Not doing `Py_INCREF` and `Py_DECREF` without reason has become ever more important I guess.

I will do that optimization; it will be very beneficial, especially for 3.12 and its immortal values. These can be used either way; they don't have to take a reference, so they can be seen as having a reference to transfer for the purposes of calling it. That will be a very nice enhancement that probably should be used generally. 

The next thing I noticed is that `signals_pending` wasn't found, and now I figured out, how the GIL build does this. The no-GIL build obviously won't do it anymore, and it's now set in bits. Thanks to work no-GIL before when it was only a fork, we already have the code change sort of how to check the new bits; I am probably going to encounter that more often. But in the CPython now, the specific function called `CONSIDER_THREADING` is, of course, something that is no longer happening.

**Comment by kayhayen** [comment]
The next thing I noticed, and it's a low hanging fruit, `f_code` in the interpreter frame got renamed to `f_executable`. We seem to only really use it for tracing purposes in Nuitka, and put it there, obviously our interest in the bytecode is very low relatively speaking.

**Comment by kayhayen** [comment]
There were plenty of uses of that code object through the frame, actually, mostly for tracebacks, outputs, etc., and often not abstracted through a function that we already had but directly. I added one for interpreter frames, too, and this seems now good, too.

The GC and the allocator, which which we deeply integrate seems to have changed, that is the next area of attack, but for today that is it. Seems like we are making good progress. Other than that, it seems mostly more freelists are affected.

**Comment by Jzhenli** [comment]
@kayhayen Hi in the coming version, if you can add a new compile option that just to compile Pure-Python packages for nuitka, since some 3rd libs have been compiled to pyd/so files by the package owner, if nuitka can skip these packages automatically, a lot of compile time will be saved.  Thanks.

**Comment by kayhayen** [comment]
@Jzhenli make new issues for unrelated things, this is not for abuse. Do not answer here too, even if I give you one, you can make a plugin that totally controls what Nuitka includes and what not. Make it, and show it to us, and maybe it can become core functionality

**Comment by kayhayen** [comment]
So, during my holiday, I made sure to do the work on the changes for the freelists. Basically, they moved it out of the interpreter and into the thread state, which kind of put on the table, that we were sort of using API calls to get at the freelist, to get the interpreter, and it turns out that not during that, and having the thread state passed in and using it from there is either equal speed (exactly actually) (Linux with libpython.so) or faster (Linux with static libpython), for tuple test cases, 0.1%, for a list test case, 0.2%, probably for dicts too, but I didn't test that. The improvements were measured for 3.12, 3.11, and 3.10.

It was nice to have the time to execute the change in a fashion that benefits all versions. I suspect a bigger enhancement is there for Windows and using the Python DLL API, but we don't measure that well yet, but I absolutely know that each API call there is terribly slow.

**Comment by kayhayen** [comment]
Then, the garbage collector integration with the allocator became a lot simpler with 3.13. It seems it is now run-time-based and only a flag is set, and it's not triggered in the allocation anymore, which spares a lot of 3.11 added and 3.12 updated code, which is now unnecessary, so that was easy.

**Comment by kayhayen** [comment]
Another area that I choose to skip until release time is uncompiled generator integration. There are a bunch of changes that are not immediately necessary for anything except async loop integration, where compiled and uncompiled generators are to be respected, and the latter doesn't have any sort of interface to call it from the outside, so we implement sending values and exceptions into it anew. That's always ugly, and a moving target, which is why I wait with that.

**Comment by kayhayen** [comment]
At that point, the C side of things compiled with Python 3.13 beta 1, and now there were linker errors. One particular noteworthy one is the removal of the asyncgen value wrapper type from API. Nuitka needs to check for it to properly integrate with uncompiled asyncgen behavior. I, however, found a function hidden behind a bytecode intrinsic table that is exposed. Calling it generates a value, from which we can grab the type.

Many more removals affected Nuitka compilation, but they were all overcome. When that happened, I was able to not only have  the hello world moment, but I could run most basic tests, even reference count tests successfully. It seems that 3.13 beta 1 has a bug with creating a temporary file for Nuitka testing, when it wants to have a temp file for a 2to3 conversion, which a few of the test cases need, and that's crashing with a permission error, but I expect that to happen only with beta1, probably a deeper Windows integration for file stuff, and not complete yet.

So, it seems we are on course for 3.13 support at the release date. Beta 2 will be released in about one week, and I will look into that as well.

**Comment by kayhayen** [comment]
Also, when looking at the free list structure, I noticed that even 3.10 exposes the dictionary objects themselves and dickeys objects, which Nuitka needs to create occasionally. My attempt at using that free list failed since I didn't check all the criteria for when it's usable, apparently for Unicode keys and smallest dict sizes only. So, that's just a new discovery of something Nuita could do, to avoid the allocator calls, and that we should do in the future too.

**Comment by kayhayen** [comment]
And regarding the asyncgen wrapper objects, we basically created a type that works the same but uses a different type value, which is not only duplicating things but requires more checks. We now check for the CPython wrapper value type and the Nuitka wrapper type, but we only need one. And also, it can only benefit if we use the same type there. So one possible cleanup/optimization is to remove our own type value there, probably only leaving the creation code in there. I think we have a freelist for those, but we can hack CPython to use our own at least for standalone mode, I need to assess how good that is, and if CPython actually ever checks the type of wrapped objects, but I believe it has to.

**Comment by reneleonhardt** [comment]
Great work 🚀
According to the docs `python -m ensurepip --upgrade` should still work
https://docs.python.org/3.13/library/ensurepip.html#command-line-interface
but there were many changes in 3.12:
https://docs.python.org/3.13/whatsnew/3.12.html#ensurepip
Maybe the beta build was incomplete in that regard 😅

**Comment by kayhayen** [comment]
Or maybe I didn't care at the time. You know, getting pip packages to work wasn't the immediate focus, I would have been happy to get closer to just following part of the API changes. When upgrading the beta2, I will check it out though.

Also, no-GIL Python will initially be incompatible with basically all extension modules not compiled for it, so it becomes its own kind of architecture. I will venture into no-GIL land when I did beta2, as you might have gathered from here, I want to demo Nuitka doing C integer looping with concurrent simple threads with 3.13 when it's released in October. It will be a benchmark only thing, meaning not translating to everything else, but as I have learned over the years, that's what most people care about only anyway ;-) but that time, it won't be a cheat, with what I mean not possible to generalize.

Until then however, I am going to focus on a bit of catchup. For 3.10 the match statement support is incomplete, even if it improved in 2.3 release now, and after that, I need to look into how 3.11 caches attribute values, and take advatange of that as well, and then operations as well, then we should be fully caught up. However, i also vowed to improve scalability of Nuitka by 10x, and that's probably going to be the largest part of it for June, hoping that beta 2 will not bring too much work, now that the noGIL Python core restructuring has been covered in Nuitka.

We shall see, but until beta 2 releases, this thread is dormant for me and the focus will largely be 2.3.1 and PyPI package catchup, as I usually do after releases.

**Comment by reneleonhardt** [comment]
All those upcoming improvements on top of an already useful compiler sound great, thank you!

**Comment by kayhayen** [comment]
So, beta2 is working and was really easy to add. I even made it part of the hotfix. 

For the next release I have started to look into the no-GIL Python variant, so far I made it compile and be detected on macOS. I am adding detection on Windows too. When it runs, in the imports, it is complaining about cyclic imports where they are not. I am assuming, I am not holding locks during import when I would have to. Even the no-GIL Python still has a GIL for some situations, e.g. module imports at least for extension modules, but maybe Python modules too.

This is looking good. As part of my 3.13 no-GIL work, I found that when making values immortal, there is a new optimization, that I am applying to 3.12 Nuitka as well. This might make GC collection faster by ignoring them by un-registering them from gc.

**Comment by kayhayen** [comment]
I was able to run the no-GIL on Windows a while ago, it was pretty easy, probably due to no extension modules being imported, the hello world kind of test worked there. I am waiting for beta3 to resume my work there.

**Comment by kayhayen** [comment]
So the release candidate should be working OK on the current develop version, I will be investigating it further this week, when I start to prepare my PyLATAM talk for no-GIL. There are reports that Linux no-GIL doesn't compile. I haven't really tried that yet, and I will add 3.13 to me CI now.

In terms of findings for optimization of 3.13 I have left that I noticed that slices in Python can be created without taking references, and Nuitka should do that too. Minor one, but I guess since 3.12 made increments/decrements more expensive, that's going to be a trend to avoid those things. And there is a dictionary keys object cache, that Nuitka fails to use properly yet. That might make it slower for some operations related to calls.

**Comment by kayhayen** [comment]
I did integrate the change I noticed for Py_DECREF to be macros with MSVC so as to not hit LTO limitations that make things slow, and I about to make the module variable and attribute variable optimization, then I will turns towards seeing how I can get extension module loading of 3.13 under control with no-GIL so standalone can be used with it.

**Comment by jimkring** [comment]
I did a quick test of Nuikta using free-threaded 3.13rc1 in Windows and the resulting executable binary (created by Nuitka) appears to have the GIL enabled. I'm wondering if I did something wrong in calling `nuitka` (like some cli argument/option) and it requires more than just running inside of a nogil python.

**Comment by kayhayen** [comment]
I am not sure how to verify that free threading is actually being used right now, a simple test would make sense. I am tempted to also separate 3.13 support, which is generally relevant, from the 3.13t support (free threading, no GIL), as the later is going to unusable for most things with standard Python I think, but I will probably do that only once I close this for 3.13 official support being a given.

**Comment by kayhayen** [comment]
So, #3062 is its own issue, I can't yet figure out, why in my testing only one core was used, and I also have questions about in-place operations in concurrent threads, where my test program locks up under pure Python. Not sure, if free threading means one needs to lock module variables for simple `+=1` operations. That will reduce its usefulness by a lot.

Anyway, the final 3.13 release seems to have made changes, that I need to follow, and I also noticed, that I forgot to do the uncompiled generator integration, which involves emulating a bunch of code and data from the core, that I only want to do for the final version. Without that, async loops are not really going to work.

That said, on factory I marked 3.13 as supported, but I will probably revert that unless I manage to get uncompiled compiler integration sorted out quickly now. Today I will be focusing on 3.13, compiling it for myself right now, and generally trying to get it out. My goal is to have 3.13 with no free threading, supported as soon as possible. I was on a lot of travels and sick 2 times in the last month, otherwise we would probably already have it, but this will be a new record anyway. But maybe 3.5 I was as quick too, and later regretted it when they changed coroutines entirely all the time. For 3.14, probably things will be even better.

**Comment by kayhayen** [comment]
So, lots of improvements for 3.13 final on factory, the rc versions are not going to work anymore and the uncompiled generator integration is sorted out too. Need to run tests to know how good it really is, I added it my CI generally. I had a few assertions from inside Python core in debug versions, hunting those down may take me this week I guess at least.

**Comment by kayhayen** [comment]
So, my CI has done plenty of 3.13, and it's looking a lot better now, but I would still have to run the CPython310 test suite, to know it's at least where stand with 3.12 so far. It will be in the 2.5 release.

**Comment by Kevin-O-Hsu** [comment]
It is a very hard work

**Comment by kayhayen** [comment]
So, unfortunately I have had to disable MinGW64 support for Windows and Python3.13, which means you will have to use MSVC, since internal structures we (illegally use) are not giving the same offsets with MSVC and MinGW64 and lead to corruption as a result. Workarounds for that will be in later releases.

**Comment by Kevin-O-Hsu** [comment]
ye but it wont be standable i think

**Comment by kayhayen** [comment]
Part of the 3.13 release I just made, but in the change log, I will have to raise cautions for using it right away. I did another fix for no-GIL 3.13, but it doesn't load extension modules, so it's relatively useless at this time and will need a lot of work before being usable over the years.

If you find your application doesn't work with 3.13 spam me senseless, but new 3.13 features are not yet even looked at really, and I got 3.12 generic types on my bucket list next. Will be running CPython test suites with 3.13 soon to see if it's good at all, and looking for minor changes not followed yet.

**Comment by kayhayen** [comment]
Please do not hi-jack other issues with questions unrelated to the issue @Eagisa

**Comment by kayhayen** [comment]
There are a few crasher issues, that make me feel a bit unhappy about the state of 3.13, unfortunately it will be a while until I can focus on those, so 3.12 is kind of still recommended.

**Comment by ArkadiuszMichalski** [comment]
@kayhayen Is there any issue where we can track when support for MinGW64 will be restored?

**Comment by kayhayen** [comment]
I will announce it in the changelog. Recently I found a big crasher source for 3.13, and am hopeful this means it should become usable. The MinGW64 issue is a bit annoying, I prepared all for it, and probably will work on it for 2.8 only, but I don't have a dedicated issue for it, nor do I want one. On my priority list however, the 3.12 generics issues are probably more important.
